File: frontend/src/lib/ai.ts
// Thin wrapper around AI endpoints on the FastAPI backend. Each function
// posts to /ai/* when a base URL is configured. Otherwise a lightweight
// on-device heuristic runs so preview features stay demonstrable.
import { http, isMockMode, cancelEvent, updateEvent, softDeleteEvent } from "./api";
import { mockApi } from "./mock-store";
import type { CareerEvent } from "./types";

export interface NLCommandPlan {
  intent: string;
  summary: string;
  actions: Array<{
    op: "update" | "cancel" | "delete" | "create";
    target_ids: string[];
    changes?: Partial<CareerEvent>;
  }>;
  affected_titles: string[];
}

export async function generateDescription(input: {
  title: string;
  company: string;
  university: string;
  industry: string;
  event_type: string;
}): Promise<string> {
  if (!isMockMode()) {
    try {
      const r = await http.post<{ description: string }>("/ai/description", input);
      return r.data.description;
    } catch { /* fall through */ }
  }
  return `Join ${input.company} at ${input.university} for an immersive ${input.event_type.replace("-", " ")} focused on ${input.industry.toLowerCase()} careers. Meet hiring managers, explore internship and graduate opportunities, and take part in on-the-spot interviews and portfolio reviews. Whether you're preparing for your first role or ready to level up, this session offers a direct line to real recruiters and practical guidance to accelerate your next career move.`;
}

export async function planNLCommand(command: string): Promise<NLCommandPlan> {
  if (!isMockMode()) {
    try {
      const r = await http.post<NLCommandPlan>("/ai/nl-command", { command });
      return r.data;
    } catch { /* fall through */ }
  }
  // Local heuristic parser (demo only).
  const c = command.toLowerCase();
  const events = mockApi.listEvents();
  const monthNames = ["january","february","march","april","may","june","july","august","september","october","november","december"];

  const stateMatch = ["kuala lumpur","selangor","penang","johor","sabah","sarawak"].find((s) => c.includes(s));
  const uniMatch = events.map((e) => e.university).find((u) => c.includes(u.toLowerCase()));
  const companyMatch = events.map((e) => e.company).find((cp) => c.includes(cp.toLowerCase()));

  const targets = events.filter((e) => {
    if (stateMatch && e.state.toLowerCase() !== stateMatch) return false;
    if (uniMatch && e.university !== uniMatch) return false;
    if (companyMatch && e.company !== companyMatch) return false;
    if (c.includes("tomorrow")) {
      const t = new Date(); t.setDate(t.getDate() + 1);
      return e.date === t.toISOString().slice(0, 10);
    }
    if (c.includes("today")) return e.date === new Date().toISOString().slice(0, 10);
    return stateMatch || uniMatch || companyMatch;
  });

  if (c.includes("cancel")) {
    return {
      intent: "cancel",
      summary: `Cancel ${targets.length} event(s)${stateMatch ? ` in ${stateMatch}` : ""}${uniMatch ? ` at ${uniMatch}` : ""}.`,
      actions: [{ op: "cancel", target_ids: targets.map((e) => e.id) }],
      affected_titles: targets.map((e) => e.title),
    };
  }
  if (c.includes("move") || c.includes("reschedule")) {
    const targetMonthIdx = monthNames.findIndex((m) => c.includes(m));
    const now = new Date();
    let newMonth = c.includes("next month") ? now.getMonth() + 1 : (targetMonthIdx >= 0 ? targetMonthIdx : now.getMonth() + 1);
    let newYear = now.getFullYear();
    if (newMonth > 11) { newMonth -= 12; newYear += 1; }
    return {
      intent: "reschedule",
      summary: `Move ${targets.length} event(s) to ${monthNames[newMonth]} ${newYear}.`,
      actions: targets.map((e) => {
        const day = e.date.slice(8, 10);
        return {
          op: "update" as const,
          target_ids: [e.id],
          changes: { date: `${newYear}-${String(newMonth + 1).padStart(2, "0")}-${day}` },
        };
      }),
      affected_titles: targets.map((e) => e.title),
    };
  }
  return {
    intent: "unknown",
    summary: `I couldn't confidently interpret that command. Try: "Cancel tomorrow's event at UTM" or "Move all Johor events to next month".`,
    actions: [],
    affected_titles: [],
  };
}

export async function applyNLPlan(plan: NLCommandPlan, user: import("./types").User | null) {
  for (const a of plan.actions) {
    for (const id of a.target_ids) {
      if (a.op === "cancel") await cancelEvent(id, user);
      else if (a.op === "update" && a.changes) await updateEvent(id, a.changes, user, "ai-updated");
      else if (a.op === "delete") await softDeleteEvent(id, user);
    }
  }
}

export interface ChatAnswer {
  answer: string;
  sources: CareerEvent[];
}
export async function ragChat(question: string): Promise<ChatAnswer> {
  if (!isMockMode()) {
    try {
      const r = await http.post<ChatAnswer>("/ai/chat", { question });
      return r.data;
    } catch { /* fall through */ }
  }
  const sources = mockApi.searchEvents(question);
  if (sources.length === 0) {
    return {
      answer: "I couldn't find any matching events. Try asking about a state, company, university or industry — for example: 'career fairs in Penang next month'.",
      sources: [],
    };
  }
  const first = sources[0];
  const list = sources
    .slice(0, 4)
    .map((e) => `• ${e.title} — ${e.university}, ${e.state} on ${e.date} (${e.start_time}–${e.end_time})`)
    .join("\n");
  return {
    answer: `I found ${sources.length} event${sources.length > 1 ? "s" : ""} that could match. Here are the top results:\n\n${list}\n\nThe closest one is "${first.title}" at ${first.location}. You can open it from the calendar to register.`,
    sources,
  };
}

export interface AiAnalytics {
  by_state: { state: string; count: number; registrations: number }[];
  by_company: { company: string; count: number; registrations: number }[];
  fill_rate: { title: string; rate: number }[];
  monthly: { month: string; events: number; registrations: number }[];
  recommendations: string[];
}
export async function aiAnalytics(): Promise<AiAnalytics> {
  if (!isMockMode()) {
    try { return (await http.get<AiAnalytics>("/ai/analytics")).data; } catch { /* fall */ }
  }
  const events = mockApi.listEvents();
  const groupBy = <K extends string>(key: (e: CareerEvent) => K) => {
    const m = new Map<K, { count: number; registrations: number }>();
    for (const e of events) {
      const k = key(e);
      const cur = m.get(k) ?? { count: 0, registrations: 0 };
      cur.count += 1; cur.registrations += e.registered_count;
      m.set(k, cur);
    }
    return m;
  };
  const byS = groupBy((e) => e.state);
  const byC = groupBy((e) => e.company);
  const monthly = groupBy((e) => e.date.slice(0, 7) as string);

  const by_state = [...byS].map(([state, v]) => ({ state, ...v })).sort((a, b) => b.registrations - a.registrations);
  const by_company = [...byC].map(([company, v]) => ({ company, ...v })).sort((a, b) => b.registrations - a.registrations);
  const fill_rate = events
    .map((e) => ({ title: e.title, rate: Math.round((e.registered_count / Math.max(e.capacity, 1)) * 100) }))
    .sort((a, b) => b.rate - a.rate)
    .slice(0, 8);
  const monthlyArr = [...monthly].map(([month, v]) => ({ month, events: v.count, registrations: v.registrations })).sort((a, b) => a.month.localeCompare(b.month));

  const top = by_state[0]; const worst = fill_rate[fill_rate.length - 1];
  const recommendations = [
    top ? `Concentrate marketing spend on ${top.state}, which drives the highest registrations (${top.registrations}).` : "Collect more data to identify top-performing regions.",
    worst ? `Consider repositioning or promoting "${worst.title}" — it has the lowest fill rate (${worst.rate}%).` : "",
    by_company[0] ? `${by_company[0].company} events consistently outperform on registrations — explore an extended partnership.` : "",
    "Bundle same-day, same-university sessions with different industries to raise cross-audience attendance.",
  ].filter(Boolean);

  return { by_state, by_company, fill_rate, monthly: monthlyArr, recommendations };
}

File: backend/app/services/ai.py
"""AI feature implementations backed by Gemini (preferred) or Ollama, with heuristic fallbacks."""

from __future__ import annotations

import json
import logging
import re
from datetime import datetime, timedelta

from sqlalchemy.orm import Session

from app.models import CareerEvent
from app.schemas import (
    ChatAnswer,
    DescriptionRequest,
    NLAction,
    NLCommandPlan,
)
from app.services import llm
from app.services.events import search_events
from app.services.llm import LLMError

logger = logging.getLogger(__name__)

# Distinct MY uni tokens — never treat as substrings of each other.
_UNI_ACRONYMS = (
    "UTeM",
    "UTHM",
    "UiTM",
    "UTM",
    "UPM",
    "UKM",
    "USM",
    "UM",
    "IIUM",
    "UIA",
    "UNITAR",
    "MMU",
    "INTI",
    "SEGi",
)

NL_PLAN_SCHEMA: dict = {
    "type": "object",
    "properties": {
        "intent": {
            "type": "string",
            "description": "One of: cancel, reschedule, unknown",
        },
        "summary": {"type": "string"},
        "actions": {
            "type": "array",
            "items": {
                "type": "object",
                "properties": {
                    "op": {
                        "type": "string",
                        "enum": ["update", "cancel", "delete", "create"],
                    },
                    "target_ids": {
                        "type": "array",
                        "items": {"type": "string"},
                    },
                    "changes": {
                        "type": "object",
                        "properties": {
                            "date": {"type": "string"},
                            "start_time": {"type": "string"},
                            "end_time": {"type": "string"},
                            "location": {"type": "string"},
                            "status": {"type": "string"},
                        },
                    },
                },
                "required": ["op", "target_ids"],
            },
        },
        "affected_titles": {
            "type": "array",
            "items": {"type": "string"},
        },
    },
    "required": ["intent", "summary", "actions", "affected_titles"],
}


def _heuristic_description(payload: DescriptionRequest) -> str:
    kind = payload.event_type.replace("-", " ")
    return (
        f"Join {payload.company} at {payload.university} for an immersive {kind} "
        f"focused on {payload.industry.lower()} careers. Meet hiring managers, explore "
        "internship and graduate opportunities, and take part in on-the-spot interviews "
        "and portfolio reviews."
    )


def generate_description(payload: DescriptionRequest) -> str:
    """Generate an event description via LLM; fall back to a template on failure."""
    prompt = (
        "Write a short professional event description for a Malaysia career-fair calendar.\n"
        "Rules: 2–3 sentences only. No markdown, no bullet points, no hashtags. "
        "Be concrete and inviting; do not invent employers, dates, or venues not given.\n\n"
        f"Title: {payload.title}\n"
        f"Company: {payload.company}\n"
        f"University: {payload.university}\n"
        f"Industry: {payload.industry}\n"
        f"Event type: {payload.event_type.replace('-', ' ')}\n"
    )
    try:
        text = llm.chat(
            [
                {
                    "role": "system",
                    "content": (
                        "You write concise career-event blurbs for Malaysian universities "
                        "and employers."
                    ),
                },
                {"role": "user", "content": prompt},
            ],
            temperature=0.5,
        )
        return text.strip().strip('"')
    except LLMError:
        logger.info("LLM unavailable for description; using heuristic")
        return _heuristic_description(payload)


def _event_catalog_line(e: CareerEvent) -> str:
    return (
        f"- id={e.id} | {e.title} | {e.company} @ {e.university} | "
        f"{e.state} | {e.date} {e.start_time}-{e.end_time} | status={e.status}"
    )


def _heuristic_nl_command(db: Session, command: str) -> NLCommandPlan:
    c = command.lower()
    events = [e for e in db.query(CareerEvent).all() if e.status not in ("cancelled", "deleted")]
    month_names = [
        "january", "february", "march", "april", "may", "june",
        "july", "august", "september", "october", "november", "december",
    ]
    states = ["kuala lumpur", "selangor", "penang", "johor", "sabah", "sarawak"]
    state_match = next((s for s in states if s in c), None)
    # Ignore blank/generic university values (e.g. "" or "all") which would
    # false-positive against common words in the command.
    uni_match = next(
        (
            e.university
            for e in events
            if e.university
            and len(e.university.strip()) > 2
            and e.university.strip().lower() not in {"all", "n/a", "na", "none"}
            and e.university.lower() in c
        ),
        None,
    )
    company_match = next(
        (
            e.company
            for e in events
            if e.company
            and len(e.company.strip()) > 2
            and e.company.lower() in c
        ),
        None,
    )

    def matches(e: CareerEvent) -> bool:
        if state_match and e.state.lower() != state_match:
            return False
        if uni_match and e.university != uni_match:
            return False
        if company_match and e.company != company_match:
            return False
        if "tomorrow" in c:
            tomorrow = (datetime.now() + timedelta(days=1)).date().isoformat()
            return e.date == tomorrow
        if "today" in c:
            return e.date == datetime.now().date().isoformat()
        return bool(state_match or uni_match or company_match)

    targets = [e for e in events if matches(e)]

    if "cancel" in c:
        return NLCommandPlan(
            intent="cancel",
            summary=f"Cancel {len(targets)} event(s).",
            actions=[NLAction(op="cancel", target_ids=[e.id for e in targets])],
            affected_titles=[e.title for e in targets],
        )

    if "move" in c or "reschedule" in c:
        now = datetime.now()
        target_idx = next((i for i, m in enumerate(month_names) if m in c), -1)
        new_month = (now.month % 12) + 1 if "next month" in c or target_idx < 0 else target_idx + 1
        new_year = now.year + (1 if new_month < now.month and "next month" in c else 0)
        if "next month" in c:
            new_month = now.month + 1
            new_year = now.year
            if new_month > 12:
                new_month = 1
                new_year += 1
        elif target_idx >= 0:
            new_month = target_idx + 1
            new_year = now.year
        actions = [
            NLAction(
                op="update",
                target_ids=[e.id],
                changes={"date": f"{new_year}-{new_month:02d}-{e.date[8:10]}"},
            )
            for e in targets
        ]
        return NLCommandPlan(
            intent="reschedule",
            summary=f"Move {len(targets)} event(s) to {month_names[new_month - 1]} {new_year}.",
            actions=actions,
            affected_titles=[e.title for e in targets],
        )

    return NLCommandPlan(
        intent="unknown",
        summary='I couldn\'t confidently interpret that command. Try: "Cancel tomorrow\'s event at UTM".',
        actions=[],
        affected_titles=[],
    )


def _validate_nl_plan(
    raw: dict,
    known_ids: set[str],
    id_to_title: dict[str, str],
    events_by_id: dict[str, CareerEvent],
    command: str,
) -> NLCommandPlan:
    """Parse LLM JSON and drop any target IDs that are not in the catalog."""
    intent = str(raw.get("intent") or "unknown").lower().strip()
    if intent not in {"cancel", "reschedule", "unknown"}:
        intent = "unknown"

    actions: list[NLAction] = []
    for item in raw.get("actions") or []:
        if not isinstance(item, dict):
            continue
        op = item.get("op")
        if op not in {"update", "cancel", "delete", "create"}:
            continue
        ids = [i for i in (item.get("target_ids") or []) if isinstance(i, str) and i in known_ids]
        ids = _filter_ids_by_command_acronyms(ids, events_by_id, command)
        if not ids:
            continue
        changes = item.get("changes") if isinstance(item.get("changes"), dict) else None
        actions.append(NLAction(op=op, target_ids=ids, changes=changes))

    affected: list[str] = []
    seen: set[str] = set()
    for action in actions:
        for tid in action.target_ids:
            title = id_to_title.get(tid)
            if title and title not in seen:
                affected.append(title)
                seen.add(title)

    summary = str(raw.get("summary") or "").strip()
    if not summary:
        if intent == "unknown" or not actions:
            summary = (
                'I couldn\'t confidently interpret that command. '
                'Try: "Cancel tomorrow\'s event at UTM".'
            )
            intent = "unknown"
            actions = []
        else:
            summary = f"Plan to {intent} {len(affected)} event(s)."

    if intent == "unknown" or not actions:
        return NLCommandPlan(
            intent="unknown",
            summary=(
                summary
                if intent == "unknown"
                else (
                    'I couldn\'t confidently interpret that command. '
                    'Try: "Cancel tomorrow\'s event at UTM".'
                )
            ),
            actions=[],
            affected_titles=[],
        )

    return NLCommandPlan(
        intent=intent,
        summary=summary,
        actions=actions,
        affected_titles=affected,
    )


def _acronyms_in_text(text: str) -> list[str]:
    found: list[str] = []
    for acr in _UNI_ACRONYMS:
        if re.search(rf"\b{re.escape(acr)}\b", text, flags=re.IGNORECASE):
            found.append(acr)
    return found


def _event_has_acronym(event: CareerEvent, acronym: str) -> bool:
    blob = f"{event.university} {event.title}"
    return bool(re.search(rf"\b{re.escape(acronym)}\b", blob, flags=re.IGNORECASE))


def _filter_ids_by_command_acronyms(
    ids: list[str],
    events_by_id: dict[str, CareerEvent],
    command: str,
) -> list[str]:
    """If the command names uni acronyms, keep only events that mention them as whole tokens."""
    mentioned = _acronyms_in_text(command)
    if not mentioned:
        return ids
    kept: list[str] = []
    for tid in ids:
        event = events_by_id.get(tid)
        if event is None:
            continue
        if any(_event_has_acronym(event, acr) for acr in mentioned):
            kept.append(tid)
    return kept


def plan_nl_command(db: Session, command: str) -> NLCommandPlan:
    """Interpret an admin NL command via LLM; fall back to keyword parsing."""
    events = [e for e in db.query(CareerEvent).all() if e.status not in ("cancelled", "deleted")]
    known_ids = {e.id for e in events}
    id_to_title = {e.id: e.title for e in events}
    events_by_id = {e.id: e for e in events}
    catalog = "\n".join(_event_catalog_line(e) for e in events) or "(no active events)"
    today = datetime.now().date().isoformat()

    user_prompt = (
        f"Today's date is {today}.\n"
        "You are an admin assistant for a Malaysia career-fair calendar.\n"
        "Interpret the admin command and return a plan. Only use event ids from the catalog.\n"
        "Supported intents: cancel, reschedule, unknown.\n"
        "For cancel: op=cancel with matching target_ids.\n"
        "For reschedule/move: op=update with changes.date as YYYY-MM-DD "
        "(keep day-of-month when only a month is mentioned, if possible).\n"
        "Matching rules:\n"
        "- Match on university, company, title, state, and date fields.\n"
        "- University acronyms are DISTINCT: UTM ≠ UTeM ≠ UTHM ≠ UM ≠ UKM ≠ USM.\n"
        "- Only match an acronym if it appears as a whole token in title or university "
        "(e.g. 'UTM' must not match 'UTeM' or 'UTHM').\n"
        "- If the command names a university/company and nothing clearly matches, "
        "return intent=unknown with empty actions.\n"
        "- If unsure, prefer unknown over guessing.\n\n"
        f"Event catalog:\n{catalog}\n\n"
        f"Admin command: {command}\n"
    )

    try:
        content = llm.chat(
            [
                {
                    "role": "system",
                    "content": (
                        "You convert natural-language admin instructions into a strict JSON plan. "
                        "Never invent event ids. Never confuse similar university acronyms. "
                        "Prefer unknown when the match is ambiguous."
                    ),
                },
                {"role": "user", "content": user_prompt},
            ],
            format=NL_PLAN_SCHEMA,
            temperature=0.0,
        )
        raw = json.loads(content)
        if not isinstance(raw, dict):
            raise ValueError("expected JSON object")
        return _validate_nl_plan(raw, known_ids, id_to_title, events_by_id, command)
    except (LLMError, json.JSONDecodeError, ValueError, TypeError) as exc:
        logger.info("LLM NL planning failed (%s); using heuristic", exc)
        return _heuristic_nl_command(db, command)


def _heuristic_rag(db: Session, question: str) -> ChatAnswer:
    sources = search_events(db, question)
    if not sources:
        return ChatAnswer(
            answer=(
                "I couldn't find any matching events. Try asking about a state, company, "
                "university or industry."
            ),
            sources=[],
        )
    lines = [
        f"• {e.title} — {e.university}, {e.state} on {e.date} ({e.start_time}–{e.end_time})"
        for e in sources[:4]
    ]
    first = sources[0]
    return ChatAnswer(
        answer=(
            f"I found {len(sources)} event(s) that could match. Here are the top results:\n\n"
            + "\n".join(lines)
            + f'\n\nThe closest one is "{first.title}" at {first.location}.'
        ),
        sources=sources,
    )


def rag_chat(db: Session, question: str) -> ChatAnswer:
    """Retrieve matching events, then ask the LLM to answer from that context only."""
    sources = search_events(db, question, limit=8)
    if not sources:
        # Broaden: if keyword search is empty, still try a short LLM apology via heuristic.
        return _heuristic_rag(db, question)

    context_lines = []
    for e in sources:
        context_lines.append(
            f"- {e.title} | {e.company} at {e.university} ({e.state}) | "
            f"{e.date} {e.start_time}-{e.end_time} | {e.event_type} | {e.industry} | "
            f"status={e.status} | location={e.location}"
        )
    context = "\n".join(context_lines)

    user_prompt = (
        "Answer the user's question about career fairs / recruitment events in Malaysia.\n"
        "Use ONLY the events listed below as your source of truth. "
        "If the list does not contain a good match, say so clearly.\n"
        "Be concise (a short paragraph plus a few bullets if helpful). "
        "Mention title, university/company, date, and location when recommending.\n\n"
        f"Retrieved events:\n{context}\n\n"
        f"User question: {question}\n"
    )

    try:
        answer = llm.chat(
            [
                {
                    "role": "system",
                    "content": (
                        "You are CareerFair's helpful assistant. Answer only from provided "
                        "event data. Do not invent events."
                    ),
                },
                {"role": "user", "content": user_prompt},
            ],
            temperature=0.3,
        )
        return ChatAnswer(answer=answer, sources=sources)
    except LLMError:
        logger.info("LLM unavailable for chat; using heuristic")
        return _heuristic_rag(db, question)


def ai_analytics(db: Session) -> dict:
    events = [e for e in db.query(CareerEvent).all() if e.status != "deleted"]

    def group(key_fn):
        m: dict[str, dict] = {}
        for e in events:
            k = key_fn(e)
            cur = m.setdefault(k, {"count": 0, "registrations": 0})
            cur["count"] += 1
            cur["registrations"] += e.registered_count
        return m

    by_s = group(lambda e: e.state)
    by_c = group(lambda e: e.company)
    monthly = group(lambda e: e.date[:7])

    by_state = sorted(
        [{"state": k, **v} for k, v in by_s.items()],
        key=lambda x: x["registrations"],
        reverse=True,
    )
    by_company = sorted(
        [{"company": k, **v} for k, v in by_c.items()],
        key=lambda x: x["registrations"],
        reverse=True,
    )
    fill_rate = sorted(
        [
            {
                "title": e.title,
                "rate": round((e.registered_count / max(e.capacity, 1)) * 100),
            }
            for e in events
        ],
        key=lambda x: x["rate"],
        reverse=True,
    )[:8]
    monthly_arr = sorted(
        [{"month": k, "events": v["count"], "registrations": v["registrations"]} for k, v in monthly.items()],
        key=lambda x: x["month"],
    )

    recommendations = []
    if by_state:
        top = by_state[0]
        recommendations.append(
            f"Concentrate marketing spend on {top['state']}, which drives the highest "
            f"registrations ({top['registrations']})."
        )
    if fill_rate:
        worst = fill_rate[-1]
        recommendations.append(
            f'Consider promoting "{worst["title"]}" — lowest fill rate ({worst["rate"]}%).'
        )
    if by_company:
        recommendations.append(
            f"{by_company[0]['company']} events outperform on registrations — explore a partnership."
        )

    return {
        "by_state": by_state,
        "by_company": by_company,
        "fill_rate": fill_rate,
        "monthly": monthly_arr,
        "recommendations": recommendations,
    }
